### ISSUE TITLE
[BUG] Bigscreen Beyond is tracked and connected, yet never wakes up display.

### ISSUE THREAD

**Issue by kzdixon** [ValveSoftware/SteamVR-for-Linux#610]
Bigscreen Beyond HMD seems to have some hurdles to overcome for working properly on Linux.  

- Bigscreen Beyond Driver is not being loaded on Linux as it should, leaving the icon as a placholder. This does not occur on Windows.
  -  Copying the contents of `~/.local/share/Steam/steamapps/common/Bigscreen Beyond Driver` to `~/.local/share/Steam/steamapps/common/SteamVR/drivers/bigscreenbeyond` remedies this.
- `lighthousedb.json` reports not being found under `bigscreenbeyond` path. This error/warning in `vrserver.txt` does not occur on Windows.
  - Remedied by copying `lighthousedb.json` under `~/.local/share/Steam/config/bigscreenbeyond`.

On Wayland with kernel 6.4.12, it appears as such via `xrandr`:
```
XWAYLAND3 disconnected (normal left inverted right x axis y axis)
   3840x1920     90.00 +
   1920x1200     90.00
   1920x1080     90.00
   1600x1200     90.00
   1680x1050     90.00
   1280x1024     90.00
   1440x900      90.00
   1280x800      90.00
   1280x720      90.00
   1024x768      90.00
   800x600       90.00
   640x480       90.00
```

After making the aforementioned tweaks to load the icons and such from the separate locations, we get to this state: 

![2023-09-14_20:32:45-screenshot](https://github.com/ValveSoftware/SteamVR-for-Linux/assets/10300293/d9133381-0483-4458-af0f-9570b5ffd2db)

The display remains black, but we can view the mirrored view and see that it shows the dashboard and tracks properly with all four lighthouses that I have.  
The best guess is that the proximity sensor either isn't working on Linux (doesn't seem to affect idle state if it times out from not moving), or that a proper display wake packet is just not being sent or being sent to the right place.  

**To Reproduce**
1. Start Linux PC with kernel 6.4.X+ (6.3.9 seems to fail to ever get it to start rendering the Steam Dashboard view or see it as a Direct Render target, it shows up as `DP-3` in `xrandr`). Leave Beyond unplugged.
2. Plug in Beyond.
3. Start SteamVR.
4. See that HMD display never wakes up as it should via proximity sensor or other movement.

**Expected behavior**
The Beyond HMD should output an image to the displays which should be visible through the lenses.

**System Information (please complete the following information):**
- Distribution: Gentoo
- SteamVR version: 1.27.5
- Steam client version: 1694466342 Mon, Sep 11 16:49 UTC -08:00 
- Opted into Steam client beta?: No
- Graphics driver version:  Mesa 23.1.7
- Gist for SteamVR System Information: https://gist.github.com/kzdixon/e36b41ea0db5f33dd735e0cbef311913

**Comment by codingcatgirl** [comment]
Hey,

super excited to see someone trying to get the beyond to work on linux and that you have made so much progress already :) Can't wait for mine to arrive.

Until then, since i'm very eager to find a solution (we apparently even both use the same graphics card)… i have a few ideas on what you could try. I had my own fair share of troubleshooting with SteamVR on linux in the last few weeks ^^;

I looked into your logs and i see several lines of "leaving standby" and "entering standby". That sounds to me like the proximity sensor works? Unless it means a different kind of standby…

If the display remains black, i can't help but wonder if that is a problem with the graphics output alltogether. I remember there being some bugs with Display Stream Compression on the newer AMD GPUs in latest linux kernels, though your version seems to be new enough so there shouldn't be any issues…

Still, i wonder if you might get different results if you set the Beyond to 72Hz (through the beyond config tool on Windows, and lets hope that will also carry over to linux then) and see if that helps. 

Otherwise i would wonder if the GPU can output to the beyond under linux at all. If it's always black, the issue might be related to that for all we know. I wonder if there's a way to add the beyond's display to the desktop as a non-direct-mode video output device and see if you get an image that way?

Also… when in doubt… even though Wayland should have the best support with SteamVR nowadays, maybe it works better on X11?

**Comment by kzdixon** [comment]
> I looked into your logs and i see several lines of "leaving standby" and "entering standby". That sounds to me like the proximity sensor works? Unless it means a different kind of standby…

Currently the only idle detection that seems to work is moving the hmd while tracked.

> If the display remains black, i can't help but wonder if that is a problem with the graphics output alltogether. I remember there being some bugs with Display Stream Compression on the newer AMD GPUs in latest linux kernels, though your version seems to be new enough so there shouldn't be any issues…

No issues on Windows in either mode. Generally when DSC was busted like with the non-official-DSC-style 144hz mode on the Index, the VR view would appear in a window on the desktop and it would complain about not being in Direct Render mode. Initial attempt was in 75hz mode before I swapped over to Windows.

**Comment by codingcatgirl** [comment]
> Currently the only idle detection that seems to work is moving the hmd while tracked.

Ah, i see. I also just saw the `Invalid input type click for controller (/user/head/proximity)` line in the logs… Still, there should at least be output then while the headset is moving i think?

> No issues on Windows in either mode. Generally when DSC was busted like with the non-official-DSC-style 144hz mode on the Index, the VR view would appear in a window on the desktop and it would complain about not being in Direct Render mode. Initial attempt was in 75hz mode before I swapped over to Windows.

Sure, but it could still be a linux-specific issue with one of those output modes.

**Comment by kzdixon** [comment]
> Ah, i see. I also just saw the Invalid input type click for controller (/user/head/proximity) line in the logs… Still, there should at least be output then while the headset is moving i think?

That line shows up in the Windows logs as well so it isn't out of the ordinary apparently for this HMD. I was told that the prox sensor isn't fully exposed to OpenXR yet as well and is basically all managed on the deivce currently for its interactions.

**Comment by codingcatgirl** [comment]
> That line shows up in the Windows logs as well so it isn't out of the ordinary apparently for this HMD. I was told that the prox sensor isn't fully exposed to OpenXR yet as well and is basically all managed on the deivce currently for its interactions.

Oh, that's unfortunate… hm.

> Well... Both don't work :(.

Could still have something to do with the output not working in general. It would be cool if you'd check if you can get it do display something in non-direct mode or if it works with x11.  Otherwise i will, but that might still take 2 months ^^

**Comment by kzdixon** [comment]
> It would be cool if you'd check if you can get it do display something in non-direct mode or if it works with x11.

I was unable to do so after changing that property. It was an idea after the initial plug-in on Windows was showing the display extended to the device until it was set to the maximum resolution. After that, SteamVR figured out what to do and properly hid it from Windows.

**Comment by stevenre3d** [comment]
I received my Beyond today. I experienced exactly the same as described in the original post: everything seemed to work out of the box, except for the part where the displays never turned on/were always black.

I'm using X11/Nvidia 3080/Manjaro.

I booted into windows, used the tool to change the refresh rate to 90, (verified that the device does work fine in windows), then booted back into linux. Now Steam-VR says "Failed to connect to headset display" with an option to restart Steam-VR.

**Comment by kzdixon** [comment]
Based on a pcap via Wireshark comparing between Win10 and Linux, it seems that this string of packets being sent to the device for full wake-up is missing:
```
810600220000
810600220000
810600220000
810600220100
810600220200
```
> These are not present in the linux pcap 
> Only present in the windows pcap
> They're all Request HID report with a response sent in top to bottom sequence each with a response HID report
> This looks like how you turn on the board inside
> The first two are sent to adjacent devices on the USB
> Last three are all sent to a single device 7.17.0
> Those last three look like the board power on for the HMD
> Total size of the packet looks to be 36byte

**Comment by TheSunCat** [comment]
Is it possible to replay these packets on Linux to wake it up manually?

**Comment by PhilT** [comment]
Has anyone managed to make any progress on this? I'd love to buy one of these but really want it to work on Linux.

**Comment by codingcatgirl** [comment]
I want to look into it myself if noone else does, but unfortunately i'm part of the European crowd and it looks like shipping there is still being delayed for a few more months at least

**Comment by PhilT** [comment]
I read on the rFactor 2 discord that shipping is actually quicker unless you have prescription lenses. I think the March date is worst case. I'm based in the UK so I hope that's the case.

I'd love to buy one and don't mind tinkering around to get it working but I'm determined to stay away from Windows now that I've moved over and need to know the BSB will work on Linux.

I look forward to your exploits!

**Comment by codingcatgirl** [comment]
> I read on the rFactor 2 discord that shipping is actually quicker unless you have prescription lenses. I think the March date is worst case. I'm based in the UK so I hope that's the case.

I recommend joining the Discord Server exclusive for people who bought a bigscreen beyond. Europe shipping has been delayed further and further because of legal requirements taking longer to resolve than expected and "late Q3" has been mentioned as the earliest realistic estimate right now.

**Comment by PhilT** [comment]
Ah, good to know.

I'd love to order one, however, until I know they work on Linux or someone tells me that they will once we all work together to solve the issues, it's a lot of money to put down on something that may or may not work on my desired platform. So far I've not booted Windows for about 6 months and I'd like to keep it that way. 

I previously had a G2 but when I switched to Linux even though some support exists in Monado I couldn't get working in my case. In theory, the BSB should be easier as it's native SteamVR. I think I'd also be okay if someone said they are working on it, don't know when it'll be working but it will eventually work then I'd probably take the plunge.

**Comment by codingcatgirl** [comment]
They are definitely not working on it themselves, and they have mentioned multiple times that it is not a priority them, often with very questionable explanations as to why. They seem to believe Linux is a difficult unsupportable territory.

Their transparency on all kinds of issues, even directly from the CEO has been absolutely amazing and makes me appreciate what they do. But then, that same CEO posts a take along the lines of "it can never be supported until NVIDIA makes better drivers and DSP is supported on linux" and i just shake my head in disbelief, given that i'm literally using my Valve Index, with DSP, on linux, right now – using a graphics card made by a small unknown company called *checks notes* AMD – and i'm not sure what they are talking about regarding NVIDIA tbh?

I think it's up to us as a community to prove them wrong and show them that one can make it work. Maybe they will pick up on it then.

**Comment by PhilT** [comment]
I think you mean DSC and as of July last year, NVIDIA added support to the Linux driver so yeah, not sure what the CEO is on about.
https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/238

Yeah I've got a NVIDIA RTX4090 and with the Index it's been pretty straightforward to setup. But I need higher res (and a lighter HMD of course!)

You keep talking, I might just place an order!

**Comment by codingcatgirl** [comment]
> You keep talking, I might just place an order!

I mean, it's refundable. The way they interact with community and provide transparency about all kinds of hickups they face on the way and now even started actively supporting and embracing community mods has made me very appreciative what they do, even if i might need to put some work myself into getting it to work on linux ^^ If you preorder and join the server you can see it for yourself.

**Comment by PhilT** [comment]
I placed a pre-order! I'll do what I can to get this going!

**Comment by yshui** [comment]
Is there a way to download the driver files without owning a Beyond?

**Comment by coolGi2007** [comment]
@yshui it should be app `2467050` on steam\
So you should be able to get it by going to steam://install/2467050

**Comment by yshui** [comment]
@coolGi2007 neat, thanks.

**Comment by yshui** [comment]
Hmm, the driver manifest says it's resource only, but it does have a `.exe`. How does SteamVR know it needs to run that exe?

**Comment by coolGi2007** [comment]
The .exe included in it is for their own utility that does stuff like change the fan speed or brightness. *Technically,* it's not required for SteamVR, but it's a nice to have.

This is the first image I found on the Discord ([message](https://discord.com/channels/816371255539138620/1151559715675983913/1196604999233450064)) for what the utility looks like
![Bigscreen Utility](https://github.com/ValveSoftware/SteamVR-for-Linux/assets/57488297/69aaf374-4d64-4687-bb31-7f2eca761ade)

Hopefully someone can make a native Linux version (as I doubt wine likes apps talking to USB devices, although I haven't checked as my unit hasn't arrived) for at least changing the brightness and refresh rate.

**Comment by TheSunCat** [comment]
If they release the source of this app, I would be very happy to buy a headset and work on porting it to Linux. The GUI is made with imgui and talking through USB is pretty doable.

**Comment by yshui** [comment]
I see, then it looks like Beyond is probably using the default lighthouse driver? Or is there a custom `driver_xxxx.so`/`driver_xxxx.dll` in `~/.local/share/Steam/steamapps/common/SteamVR/drivers/bigscreenbeyond`?

**Comment by Scrumplex** [comment]
> I see, then it looks like Beyond is probably using the default lighthouse driver? 

Indeed!
The main blocker right now are subtle differences in the lighthouse driver between Windows and Linux.

See https://github.com/ValveSoftware/SteamVR-for-Linux/issues/610#issuecomment-1761905276

**Comment by PhilT** [comment]
I created a Discord server where we can discuss Bigscreen Beyond on Linux. I thought it made sense to have a place to collaborate in while we try to get this thing working on Linux. Join me here: https://discord.gg/ztgm4eX72E

**Comment by c-x-berger** [comment]
@kzdixon, can you provide that pcap you mentioned? It could be really useful to some of the other folks trying to work on this.

**Comment by BrianNormant** [comment]
> Based on a pcap via Wireshark comparing between Win10 and Linux, it seems that this string of packets being sent to the device for full wake-up is missing:
> 
> ```
> 810600220000
> 810600220000
> 810600220000
> 810600220100
> 810600220200
> ```
> 
> > These are not present in the linux pcap
> > Only present in the windows pcap
> > They're all Request HID report with a response sent in top to bottom sequence each with a response HID report
> > This looks like how you turn on the board inside
> > The first two are sent to adjacent devices on the USB
> > Last three are all sent to a single device 7.17.0
> > Those last three look like the board power on for the HMD
> > Total size of the packet looks to be 36byte

Could you like the pngcap of the specific packets.
I'm trying to manually send those via with hidapi, but I can't understand what part of the usb packet those numbers represents.

**Comment by c-x-berger** [comment]
Just as a side note / update for people watching this ticket who aren't wired directly into three different Discord servers, Bigscreen claims Valve needs to reach out to them to fix this. (I'm not sure if this means Bigscreen knows what's wrong with the driver and is just waiting for Valve to ask or what.)

![Screenshot_20240228-032748.png](https://github.com/ValveSoftware/SteamVR-for-Linux/assets/43154879/6b520537-2338-4d85-9588-0871d45fcdb8)

There hasn't been much success with trying to reverse engineer things as outsiders.

**Comment by kzdixon** [comment]
The Bigscreen Beyond appears to work finally (display giving visible picture) on kernel 6.7.6 (zen in my case) with the following patches that have yet to be upstreamed:

- https://lkml.org/lkml/2022/2/20/153
- https://lkml.org/lkml/2022/2/20/151

Note: At least one of these patches had to be updated for the zen codebase I was using from zen-sources as of 6.7.5+.

Optionally, an EDID Quirk patch can be used just in case, but may not be necessary:
```diff
diff --git a/drivers/gpu/drm/drm_edid.c b/drivers/gpu/drm/drm_edid.c
index 3b4065099..639699e3b 100644
--- a/drivers/gpu/drm/drm_edid.c
+++ b/drivers/gpu/drm/drm_edid.c
@@ -189,6 +189,9 @@ static const struct edid_quirk {
        /* Rotel RSX-1058 forwards sink's EDID but only does HDMI 1.1*/
        EDID_QUIRK('E', 'T', 'R', 13896, EDID_QUIRK_FORCE_8BPC),

+       /* Bigscreen Beyond Headset */
+       EDID_QUIRK('B', 'I', 'G', 0x1234, EDID_QUIRK_NON_DESKTOP),
+
        /* Valve Index Headset */
        EDID_QUIRK('V', 'L', 'V', 0x91a8, EDID_QUIRK_NON_DESKTOP),
        EDID_QUIRK('V', 'L', 'V', 0x91b0, EDID_QUIRK_NON_DESKTOP),

```
The Beyond has been observed working on both Linux SteamVR 2.3.5 (1.27 should work as well I assume) and Monado via Envision.

![2024-02-28-20:43:08-screenshot](https://github.com/ValveSoftware/SteamVR-for-Linux/assets/10300293/780cb89b-1ae5-4ba2-8d4b-4638e9af6b7b)
![2024-02-28-16:23:43-screenshot](https://github.com/ValveSoftware/SteamVR-for-Linux/assets/10300293/67ef76b8-4887-444e-bc78-0c0b6e752d75)

As soon as those DSC patches get upstreamed (they were taken on a whim from https://github.com/santeri3700/vive-pro-2-on-linux/blob/master/KERNEL.md since DSC is involved with the Beyond as well), the Beyond should theoretically start working out of the box for users, though edits/updates to udev rules may be necessary for hidraw (not verified).

In theory this may prove that it was never a SteamVR issue after all but at the very least this gave us a spot to track the issue.

**Comment by SpookySkeletons** [comment]
The EDID quirk is 100% necessary, upstream it when you can.

Wayland has no means to mark a display non-desktop && not many want to have to do this by hand each login.

**Comment by DaeTheDragon** [comment]
> The Bigscreen Beyond appears to work finally (display giving visible picture) on kernel 6.7.6 (zen in my case) with the following patches that have yet to be upstreamed:

Hiya, as someone who doesn't have experience with patching kernels, how would I go about doing this?
I'm running arch if that makes a difference
I tried to look up tutorials for patching the Linux kernel, but they seem to be telling me to use a patch file of some kind, whereas the linked sites provided seem to just have text content, so I'm really not sure where to start.
I'm buusy for the next few days but advice would be appreciated!

**Comment by PhilT** [comment]
> > The Bigscreen Beyond appears to work finally (display giving visible picture) on kernel 6.7.6 (zen in my case) with the following patches that have yet to be upstreamed:
> 
> Hiya, as someone who doesn't have experience with patching kernels, how would I go about doing this? I'm running arch if that makes a difference I tried to look up tutorials for patching the Linux kernel, but they seem to be telling me to use a patch file of some kind, whereas the linked sites provided seem to just have text content, so I'm really not sure where to start. I'm buusy for the next few days but advice would be appreciated!

Pop over to this Discord: https://discord.gg/ztgm4eX72E or this one: https://discord.gg/CTmD94Vw when you're ready and someone will be able to help you.

**Comment by CertainLach** [comment]
> Edit: These patches will presumably allow the vive pro 2 to function and possibly the upcoming somnium VR1 and much more exotic HMDs that are lighthouse based.

Vive Pro 2 will not work OOB due to nonstandard lens calibration config, In my Vive Pro 2 driver for Linux (https://github.com/CertainLach/VivePro2-Linux-Driver/) I had to use wine to call Vive provided calibration libraries directly.

Good to hear that my dsc bits-per-pixel patch can be useful for at least two VR headsets, maybe it is the time I will finally update and upstream it.

**Comment by CertainLach** [comment]
> Wayland has no means to mark a display non-desktop && not many want to have to do this by hand each

Actually, you can replace headset EDID to provide display use-case as VR, Vive Pro 2 does that, and kernel respects this EDID section now, without need to rely on quirks.

**Comment by Teteros** [comment]
Talking to the users on various discords it seems currently BSB is working well on AMD but haven't heard any feedback from anyone else on NVIDIA.

On NVIDIA, the non-desktop property is detected and set out of the box so the EDID patches aren't needed and the HMD display can wake up fine on SteamVR/Monado. However at least on my setup (Arch Linux, RTX 4090, proprietary nvidia-dkms 550.90 drivers) the BSB is stuck on a gray screen (which normally would indicate lighthouse tracking fail, though they work fine on Windows) and there's this rectangular plane of fuzzy static noise on the top half of both eyes that doesn't go away whether you're looking at SteamVR Home or demo apps like xrgears etc.  

I took a badly lit picture of it here 
![image](https://github.com/ValveSoftware/SteamVR-for-Linux/assets/5067989/49259a3d-17be-4a95-8aeb-322fad339600)
noise pattern looks like:
![image](https://github.com/ValveSoftware/SteamVR-for-Linux/assets/5067989/849d3a22-99bf-4a5a-86d0-0524d1e20fed)


This was on the 90hz mode, I haven't tested 75hz yet.

**Comment by Hisanatos** [comment]
From what I was able to find, on Nvidia it seems to be a general issue, which either requires someone to look into the Open Kernel from Nvidia, or Nvidia themselves fixing it on their side. https://forums.developer.nvidia.com/t/bigscreen-beyond-hmd-unable-to-be-initialized-on-nvidia-possibly-due-to-dsc-bpp-issue/294108

But if it works on AMD, that means the problem is definetly on the GPU driver side not somewhere else.

**Comment by SpookySkeletons** [comment]
@Hisanatos I submitted that nvidia issue but I do not have an nvidia GPU nor a BSB to help debug this issue

All available in this thread please make an account on nvidia's issue tracker and relocate this problem to help nvidia address. Their rep needs more info on the issue ticket, please do send what you can.

**Comment by HcgRandon** [comment]
@SpookySkeletons Why nvidia? I am experiencing this issue on my AMD GPU

**Comment by adamredwine** [comment]
I'm having the same issue with BSB and an 3080ti.  I'm going to try to tinker around with arch and the suggested patches.

**Comment by kzdixon** [comment]
> I'm having the same issue with BSB and an 3080ti. I'm going to try to tinker around with arch and the suggested patches.

You'll be waiting on Nvidia to add BPP support to their DSC implementation on Linux. Last info was "months away".

https://forums.developer.nvidia.com/t/bigscreen-beyond-hmd-unable-to-be-initialized-on-nvidia-possibly-due-to-dsc-bpp-issue/294108/16

**Comment by pelasgus** [comment]
@HcgRandon The aforementioned patches provided here by @kzdixon were meant to solve the issue for AMD GPUs; however it seems that the patch requires updating to be made compatible with the latest kernel.

**Comment by kzdixon** [comment]
> [@HcgRandon](https://github.com/HcgRandon) The aforementioned patches provided here by [@kzdixon](https://github.com/kzdixon) were meant to solve the issue for AMD GPUs; however it seems that the patch requires updating to be made compatible with the latest kernel.

There's an updated patchset for 6.10 that works up until 6.15 here: https://gist.githubusercontent.com/galister/08cddf10ac18929647d5fb6308df3e4b/raw/0f6417b6cb069f19d6c28b730499c07de06ec413/combined-bsb-6-10.patch (from https://lvra.gitlab.io/docs/hardware/)

6.15 required a few more changes (borrowed from another user):
<details>

```diff
diff --git a/drivers/gpu/drm/drm_edid.c b/drivers/gpu/drm/drm_edid.c
index 3b4065099..639699e3b 100644
--- a/drivers/gpu/drm/drm_edid.c
+++ b/drivers/gpu/drm/drm_edid.c
@@ -189,6 +189,9 @@ static const struct edid_quirk {
 	/* Rotel RSX-1058 forwards sink's EDID but only does HDMI 1.1*/
 	EDID_QUIRK('E', 'T', 'R', 13896, EDID_QUIRK_FORCE_8BPC),
 
+	/* Bigscreen Beyond Headset */
+	EDID_QUIRK('B', 'I', 'G', 0x1234, EDID_QUIRK_NON_DESKTOP),
+
 	/* Valve Index Headset */
 	EDID_QUIRK('V', 'L', 'V', 0x91a8, EDID_QUIRK_NON_DESKTOP),
 	EDID_QUIRK('V', 'L', 'V', 0x91b0, EDID_QUIRK_NON_DESKTOP),

From c33583995576e9ac532c4ad9e260324b1c4fa3a3 Mon Sep 17 00:00:00 2001
From: Yaroslav Bolyukin <iam@lach.pw>
Date: Sun, 30 Oct 2022 19:04:26 +0100
Subject: [PATCH 3/3] drm/amd: use fixed dsc bits-per-pixel from edid

VESA vendor header from DisplayID spec may contain fixed bit per pixel
rate, it should be respected by drm driver

Signed-off-by: Yaroslav Bolyukin <iam@lach.pw>
Reviewed-by: Wayne Lin <Wayne.Lin@amd.com>
---
 drivers/gpu/drm/amd/display/dc/core/dc_stream.c           | 2 ++
 drivers/gpu/drm/amd/display/dc/dc_types.h                 | 3 +++
 2 files changed, 5 insertions(+)

diff --git a/drivers/gpu/drm/amd/display/dc/core/dc_stream.c b/drivers/gpu/drm/amd/display/dc/core/dc_stream.c
index 38d71b5c1f2d..f2467b64268b 100644
--- a/drivers/gpu/drm/amd/display/dc/core/dc_stream.c
+++ b/drivers/gpu/drm/amd/display/dc/core/dc_stream.c
@@ -103,6 +103,8 @@ static bool dc_stream_construct(struct dc_stream_state *stream,
 
 	/* EDID CAP translation for HDMI 2.0 */
 	stream->timing.flags.LTE_340MCSC_SCRAMBLE = dc_sink_data->edid_caps.lte_340mcsc_scramble;
+	stream->timing.dsc_fixed_bits_per_pixel_x16 =
+		dc_sink_data->edid_caps.dsc_fixed_bits_per_pixel_x16;
 
 	memset(&stream->timing.dsc_cfg, 0, sizeof(stream->timing.dsc_cfg));
 	stream->timing.dsc_cfg.num_slices_h = 0;
diff --git a/drivers/gpu/drm/amd/display/dc/dc_types.h b/drivers/gpu/drm/amd/display/dc/dc_types.h
index dc78e2404b48..65915a10ab48 100644
--- a/drivers/gpu/drm/amd/display/dc/dc_types.h
+++ b/drivers/gpu/drm/amd/display/dc/dc_types.h
@@ -231,6 +231,9 @@ struct dc_edid_caps {
 	bool edid_hdmi;
 	bool hdr_supported;
 
+	/* DisplayPort caps */
+	uint32_t dsc_fixed_bits_per_pixel_x16;
+
 	struct dc_panel_patch panel_patch;
 };
 
-- 
2.38.1


diff --git a/drivers/gpu/drm/drm_edid.c b/drivers/gpu/drm/drm_edid.c
index 3b4065099..15268afa3 100644
--- a/drivers/gpu/drm/drm_edid.c
+++ b/drivers/gpu/drm/drm_edid.c
@@ -6391,7 +6391,7 @@ static void drm_parse_vesa_mso_data(struct drm_connector *connector,
 	if (oui(vesa->oui[0], vesa->oui[1], vesa->oui[2]) != VESA_IEEE_OUI)
 		return;
 
-	if (sizeof(*vesa) != sizeof(*block) + block->num_bytes) {
+	if (block->num_bytes < 5) {
 		drm_dbg_kms(connector->dev,
 			    "[CONNECTOR:%d:%s] Unexpected VESA vendor block size\n",
 			    connector->base.id, connector->name);
@@ -6414,24 +6414,37 @@ static void drm_parse_vesa_mso_data(struct drm_connector *connector,
 		break;
 	}
 
-	if (!info->mso_stream_count) {
-		info->mso_pixel_overlap = 0;
-		return;
-	}
+	info->mso_pixel_overlap = 0;
+
+	if (info->mso_stream_count) {
+		info->mso_pixel_overlap = FIELD_GET(DISPLAYID_VESA_MSO_OVERLAP, vesa->mso);
+
+		if (info->mso_pixel_overlap > 8) {
+			drm_dbg_kms(connector->dev,
+					"[CONNECTOR:%d:%s] Reserved MSO pixel overlap value %u\n",
+					connector->base.id, connector->name,
+					info->mso_pixel_overlap);
+			info->mso_pixel_overlap = 8;
+		}
 
-	info->mso_pixel_overlap = FIELD_GET(DISPLAYID_VESA_MSO_OVERLAP, vesa->mso);
-	if (info->mso_pixel_overlap > 8) {
 		drm_dbg_kms(connector->dev,
-			    "[CONNECTOR:%d:%s] Reserved MSO pixel overlap value %u\n",
-			    connector->base.id, connector->name,
-			    info->mso_pixel_overlap);
-		info->mso_pixel_overlap = 8;
+				"[CONNECTOR:%d:%s] MSO stream count %u, pixel overlap %u\n",
+				connector->base.id, connector->name,
+				info->mso_stream_count, info->mso_pixel_overlap);
+	}
+
+	if (block->num_bytes < 7) {
+		/* DSC bpp is optional */
+		return;
 	}
 
+	info->dp_dsc_bpp = FIELD_GET(DISPLAYID_VESA_DSC_BPP_INT, vesa->dsc_bpp_int) * 16 +
+		FIELD_GET(DISPLAYID_VESA_DSC_BPP_FRACT, vesa->dsc_bpp_fract);
+
 	drm_dbg_kms(connector->dev,
-		    "[CONNECTOR:%d:%s] MSO stream count %u, pixel overlap %u\n",
-		    connector->base.id, connector->name,
-		    info->mso_stream_count, info->mso_pixel_overlap);
+		"[CONNECTOR:%d:%s] DSC bits per pixel %u\n",
+		connector->base.id, connector->name,
+		info->dp_dsc_bpp);
 }
 
 static void drm_update_mso(struct drm_connector *connector,
@@ -6479,6 +6492,7 @@ static void drm_reset_display_info(struct drm_connector *connector)
 	info->mso_stream_count = 0;
 	info->mso_pixel_overlap = 0;
 	info->max_dsc_bpp = 0;
+	info->dp_dsc_bpp = 0;
 
 	kfree(info->vics);
 	info->vics = NULL;
diff --git a/drivers/gpu/drm/drm_edid.c.rej b/drivers/gpu/drm/drm_edid.c.rej
new file mode 100644
index 000000000..de3b3bf4e
--- /dev/null
+++ b/drivers/gpu/drm/drm_edid.c.rej
@@ -0,0 +1,9 @@
+diff a/drivers/gpu/drm/drm_edid.c b/drivers/gpu/drm/drm_edid.c	(rejected hunks)
+@@ -6376,6 +6389,7 @@ static void drm_reset_display_info(struct drm_connector *connector)
+ 	info->mso_stream_count = 0;
+ 	info->mso_pixel_overlap = 0;
+ 	info->max_dsc_bpp = 0;
++	info->dp_dsc_bpp = 0;
+ }
+ 
+ static u32 update_display_info(struct drm_connector *connector,
diff --git a/include/drm/drm_connector.h b/include/drm/drm_connector.h
index fe88d7fc6..1de1d1726 100644
--- a/include/drm/drm_connector.h
+++ b/include/drm/drm_connector.h
@@ -803,6 +803,12 @@ struct drm_display_info {
 	 */
 	u32 max_dsc_bpp;
 
+	/**
+	 * @dp_dsc_bpp: DP Display-Stream-Compression (DSC) timing's target
+	 * DST bits per pixel in 6.4 fixed point format. 0 means undefined
+	 */
+	u16 dp_dsc_bpp;
+
 	/**
 	 * @vics: Array of vics_len VICs. Internal to EDID parsing.
 	 */
diff --git a/include/drm/drm_connector.h.rej b/include/drm/drm_connector.h.rej
new file mode 100644
index 000000000..d54d40443
--- /dev/null
+++ b/include/drm/drm_connector.h.rej
@@ -0,0 +1,13 @@
+diff a/include/drm/drm_connector.h b/include/drm/drm_connector.h	(rejected hunks)
+@@ -721,6 +721,11 @@ struct drm_display_info {
+ 	 * monitor's default value is used instead.
+ 	 */
+ 	u32 max_dsc_bpp;
++	/**
++	 * @dp_dsc_bpp: DP Display-Stream-Compression (DSC) timing's target
++	 * DST bits per pixel in 6.4 fixed point format. 0 means undefined
++	 */
++	u16 dp_dsc_bpp;
+ };
+ 
+ int drm_display_info_set_bus_formats(struct drm_display_info *info,
diff --git a/include/drm/drm_displayid.h b/include/drm/drm_displayid.h
index 566497eeb..3a4bd0816 100644
--- a/drivers/gpu/drm/drm_displayid_internal.h
+++ b/drivers/gpu/drm/drm_displayid_internal.h
@@ -131,12 +131,16 @@ struct displayid_detailed_timing_block {
 
 #define DISPLAYID_VESA_MSO_OVERLAP	GENMASK(3, 0)
 #define DISPLAYID_VESA_MSO_MODE		GENMASK(6, 5)
+#define DISPLAYID_VESA_DSC_BPP_INT	GENMASK(5, 0)
+#define DISPLAYID_VESA_DSC_BPP_FRACT GENMASK(3, 0)
 
 struct displayid_vesa_vendor_specific_block {
 	struct displayid_block base;
 	u8 oui[3];
 	u8 data_structure_type;
 	u8 mso;
+	u8 dsc_bpp_int;
+	u8 dsc_bpp_fract;
 } __packed;
 
 /*
diff --git a/include/drm/drm_displayid.h.rej b/include/drm/drm_displayid.h.rej
new file mode 100644
index 000000000..61fbd38e0
--- /dev/null
+++ b/include/drm/drm_displayid.h.rej
@@ -0,0 +1,18 @@
+diff a/include/drm/drm_displayid.h b/include/drm/drm_displayid.h	(rejected hunks)
+@@ -131,12 +131,16 @@ struct displayid_detailed_timing_block {
+ 
+ #define DISPLAYID_VESA_MSO_OVERLAP	GENMASK(3, 0)
+ #define DISPLAYID_VESA_MSO_MODE		GENMASK(6, 5)
++#define DISPLAYID_VESA_DSC_BPP_INT	GENMASK(5, 0)
++#define DISPLAYID_VESA_DSC_BPP_FRACT GENMASK(3, 0)
+ 
+ struct displayid_vesa_vendor_specific_block {
+ 	struct displayid_block base;
+ 	u8 oui[3];
+ 	u8 data_structure_type;
+ 	u8 mso;
++	u8 dsc_bpp_int;
++	u8 dsc_bpp_fract;
+ } __packed;
+ 
+ /* DisplayID iteration */

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_helpers.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_helpers.c
index d4395b92fb85..6c7f589e19ac 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_helpers.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_helpers.c
@@ -136,6 +136,8 @@ enum dc_edid_status dm_helpers_parse_edid_caps(
 
 	edid_caps->edid_hdmi = connector->display_info.is_hdmi;
 
+	edid_caps->dsc_fixed_bits_per_pixel_x16 = connector->display_info.dp_dsc_bpp;
+	
 	apply_edid_quirks(dev, edid_buf, edid_caps);
 
 	sad_count = drm_edid_to_sad((struct edid *) edid->raw_edid, &sads);
```

</details>

**Comment by pelasgus** [comment]
> [!TIP]
>  Create a patch file and copy the contents of the relative patch from the aforementioned comments.
> For NixOS users on standard, simply import the following file:
```nix
{config, pkgs, lib,  ...}:
{
  boot.kernelPackages = pkgs.linuxPackages;
  # Bigscreen Beyond Patch
  # https://github.com/ValveSoftware/SteamVR-for-Linux/issues/610
  boot.kernelPatches = [
    {
      name = "dsc";
      patch = ./patches/vesa-dsc-fixed-bpp.patch;
    }
  ];
}
```

**Comment by razzle** [comment]
> > [@HcgRandon](https://github.com/HcgRandon) The aforementioned patches provided here by [@kzdixon](https://github.com/kzdixon) were meant to solve the issue for AMD GPUs; however it seems that the patch requires updating to be made compatible with the latest kernel.
> 
> There's an updated patchset for 6.10 that works up until 6.15 here: https://gist.githubusercontent.com/galister/08cddf10ac18929647d5fb6308df3e4b/raw/0f6417b6cb069f19d6c28b730499c07de06ec413/combined-bsb-6-10.patch (from https://lvra.gitlab.io/docs/hardware/)
> 
> 6.15 required a few more changes (borrowed from another user):

I can confirm that this patch set also makes the Beyond 2 work and applies to 6.16 as well.

**Comment by dotbanana** [comment]
Thanks for the information. 
There is also a patch for >6.15 as referenced here:  
https://lvra.gitlab.io/docs/hardware/  
Link: https://lvra.gitlab.io/docs/hardware/bigscreen-beyond-kernel-6.15.patch  

Is it possible that this will get by default into a future kernel update? I have no experience with those things, i am a simple user who wants to use the BSB2 on Linux.

**Comment by coolGi69** [comment]
Hey, I'm the person who ported the original patch to 6.15+. 

> Is it possible that this will get by default into a future kernel update? I have no experience with those things, i am a simple user who wants to use the BSB2 on Linux.

The patch doesn't often change, so you shouldn't need to worry about it (as you can tell by this being the only patch to the patch since the original), just put in whatever thing you have for your distro to auto update the kernel with the patch.
When it does break, you can check back on the LVRA hardware page.

**Comment by dotbanana** [comment]
> just put in whatever thing you have for your distro to auto update the kernel with the patch  

I am using Fedora and there isn't such thing by what i have found out in my research. But again i am mostly a normal user, i am not experienced in messing with kernel and bootloaders and all those things. 
I made a feature request for the kernel patch being included in those custom Fedora-Kernels here and the person maintaining it was so nice to implement it immediatly:
https://github.com/whitehara/kernel-patch-fedora/issues/2  

So for me this is solved that way. Maybe this helps also other folks :)

**Comment by coolGi69** [comment]
# Notice for people waiting on this and only tuned in here
Just in case anyone is only tuned into this thread and wondering the state of the BSB for Linux, then just making a PSA that it is in a stable state here on Linux.
The new kernel patches are over on the [Linux VR Adventures wiki Bigscreen Beyond page](https://wiki.vronlinux.org/docs/hardware/bigscreen-beyond/), with an effort to upstream into mainline Linux. The previously mentioned DCS issues (rainbow static) have also been fixed through kernel patches, which is also available on the same LVRA wiki page. Most future discussions and support are currently having on the LVRA Discord/Matrix which you can find [here](https://wiki.vronlinux.org/docs/community/) (with a dedicated BSB channel/room for any discussions). Libsurvive with the BSB has also been fixed for anyone who cares about that.
All these fixes apply to both SteamVR & Monado (the preferred/FOSS VR runtime).

Anyways, hope this helps anyone who just found this issue out of the blue, happy Linuxing ^^
